### ISSUE TITLE
Python crash when running as PCP PMDA plugin

### ISSUE THREAD

**Issue by myllynen** [iovisor/bcc#1533]
While testing PCP BCC PMDA (see https://github.com/performancecopilot/pcp/pull/409) we occasionally see Python that runs the BCC module crashing on pmcd (the PCP daemon running plugins) shutdown. PCP has other Python plugins as well and we've never seen this issue with them so this and the stack trace make me think the issue is originating from BCC. I realize this might be a hard-to-reproduce issue especially for those not familiar with PCP but the initial GitHub page at https://github.com/myllynen/pcp-bcc-pmda provides step-by-step instructions how to set this is up and I'd be happy to test any possible patch proposals if someone sees what might be causing this.

```
Process 6993 (python3) of user 0 dumped core.

Stack trace of thread 7008:
#0  0x00007f4c142916cd perf_reader_event_read (libbcc.so.0)
#1  0x00007f4c14291b03 perf_reader_poll (libbcc.so.0)
#2  0x00007f4c18979d1e ffi_call_unix64 (libffi.so.6)
#3  0x00007f4c1897968f ffi_call (libffi.so.6)
#4  0x00007f4c18b8da29 _call_function_pointer (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#5  0x00007f4c18b8dc14 PyCFuncPtr_call (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#6  0x00007f4c22058204 _PyObject_FastCallDict (libpython3.6m.so.1.0)
#7  0x00007f4c220c73ad call_function (libpython3.6m.so.1.0)
#8  0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#9  0x00007f4c220570b3 _PyEval_EvalCodeWithName (libpython3.6m.so.1.0)
#10 0x00007f4c22089e11 fast_function (libpython3.6m.so.1.0)
#11 0x00007f4c220c733e call_function (libpython3.6m.so.1.0)
#12 0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#13 0x00007f4c22057d9a _PyFunction_FastCall (libpython3.6m.so.1.0)
#14 0x00007f4c220583de _PyObject_FastCallDict (libpython3.6m.so.1.0)
#15 0x00007f4c22062261 _PyObject_Call_Prepend (libpython3.6m.so.1.0)
#16 0x00007f4c220587fb PyObject_Call (libpython3.6m.so.1.0)
#17 0x00007f4c220f28b7 do_call_core (libpython3.6m.so.1.0)
#18 0x00007f4c22089bea _PyFunction_FastCall (libpython3.6m.so.1.0)
#19 0x00007f4c220c733e call_function (libpython3.6m.so.1.0)
#20 0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#21 0x00007f4c22089bea _PyFunction_FastCall (libpython3.6m.so.1.0)
#22 0x00007f4c220c733e call_function (libpython3.6m.so.1.0)
#23 0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#24 0x00007f4c22057d9a _PyFunction_FastCall (libpython3.6m.so.1.0)
#25 0x00007f4c220583de _PyObject_FastCallDict (libpython3.6m.so.1.0)
#26 0x00007f4c22062261 _PyObject_Call_Prepend (libpython3.6m.so.1.0)
#27 0x00007f4c220587fb PyObject_Call (libpython3.6m.so.1.0)
#28 0x00007f4c2215fd02 t_bootstrap (libpython3.6m.so.1.0)
#29 0x00007f4c21d2261b start_thread (libpthread.so.0)
#30 0x00007f4c212f391f __clone (libc.so.6)

Stack trace of thread 6993:
#0  0x00007f4c21d2c56b __write (libpthread.so.0)
#1  0x00007f4c1428f0f5 bpf_detach_probe (libbcc.so.0)
#2  0x00007f4c18979d1e ffi_call_unix64 (libffi.so.6)
#3  0x00007f4c1897968f ffi_call (libffi.so.6)
#4  0x00007f4c18b8da29 _call_function_pointer (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#5  0x00007f4c18b8dc14 PyCFuncPtr_call (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#6  0x00007f4c22058204 _PyObject_FastCallDict (libpython3.6m.so.1.0)
#7  0x00007f4c220c73ad call_function (libpython3.6m.so.1.0)
#8  0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#9  0x00007f4c22089bea _PyFunction_FastCall (libpython3.6m.so.1.0)
#10 0x00007f4c220c733e call_function (libpython3.6m.so.1.0)
#11 0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#12 0x00007f4c22089bea _PyFunction_FastCall (libpython3.6m.so.1.0)
#13 0x00007f4c220c733e call_function (libpython3.6m.so.1.0)
#14 0x00007f4c220f0eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#15 0x00007f4c220d07e7 _PyEval_EvalCodeWithName (libpython3.6m.so.1.0)
#16 0x00007f4c220d1223 function_call (libpython3.6m.so.1.0)
#17 0x00007f4c220587fb PyObject_Call (libpython3.6m.so.1.0)
#18 0x00007f4c22147db1 atexit_callfuncs (libpython3.6m.so.1.0)
#19 0x00007f4c2215ef5a call_py_exitfuncs (libpython3.6m.so.1.0)
#20 0x00007f4c22160142 Py_Main (libpython3.6m.so.1.0)
#21 0x000055b101a71d45 main (python3.6)
#22 0x00007f4c211fd00a __libc_start_main (libc.so.6)
#23 0x000055b101a71eba _start (python3.6)
Process 6993 (python3.6) crashed in perf_reader_event_read()
```

Thanks.

**Comment by yonghong-song** [comment]
Thanks for reporting this issue. I had a theory why this could happen and crafted the below patch,
```
-bash-4.2$ git diff
diff --git a/src/cc/perf_reader.c b/src/cc/perf_reader.c
index 6054a8b..2693a4f 100644
--- a/src/cc/perf_reader.c
+++ b/src/cc/perf_reader.c
@@ -266,7 +266,7 @@ void perf_reader_event_read(struct perf_reader *reader) {
       fprintf(stderr, "%s: unknown sample type %d\n", __FUNCTION__, e->type);
     }
 
-    write_data_tail(perf_header, perf_header->data_tail + e->size);
+    write_data_tail(perf_header, (perf_header->data_tail + e->size) % buffer_size);
   }
 }
```

Could you help give a try? Thanks!

**Comment by myllynen** [comment]
Thanks for your prompt reply! I've now upgraded to 0.5.0 (on Fedora 27), recompiled the Fedora RPM locally, and verified seeing the issue (as expected). Then I applied this patch and recompiled again but I still see the same trace:

```
Process 14072 (python3) of user 0 dumped core.

Stack trace of thread 14087:
#0  0x00007f9612aeecad perf_reader_event_read (libbcc.so.0)
#1  0x00007f9612aef0f3 perf_reader_poll (libbcc.so.0)
#2  0x00007f961742cd1e ffi_call_unix64 (libffi.so.6)
#3  0x00007f961742c68f ffi_call (libffi.so.6)
#4  0x00007f9617640a29 _call_function_pointer (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#5  0x00007f9617640c14 PyCFuncPtr_call (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#6  0x00007f9620b0b204 _PyObject_FastCallDict (libpython3.6m.so.1.0)
#7  0x00007f9620b7a3ad call_function (libpython3.6m.so.1.0)
#8  0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#9  0x00007f9620b0a0b3 _PyEval_EvalCodeWithName (libpython3.6m.so.1.0)
#10 0x00007f9620b3ce11 fast_function (libpython3.6m.so.1.0)
#11 0x00007f9620b7a33e call_function (libpython3.6m.so.1.0)
#12 0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#13 0x00007f9620b0ad9a _PyFunction_FastCall (libpython3.6m.so.1.0)
#14 0x00007f9620b0b3de _PyObject_FastCallDict (libpython3.6m.so.1.0)
#15 0x00007f9620b15261 _PyObject_Call_Prepend (libpython3.6m.so.1.0)
#16 0x00007f9620b0b7fb PyObject_Call (libpython3.6m.so.1.0)
#17 0x00007f9620ba58b7 do_call_core (libpython3.6m.so.1.0)
#18 0x00007f9620b3cbea _PyFunction_FastCall (libpython3.6m.so.1.0)
#19 0x00007f9620b7a33e call_function (libpython3.6m.so.1.0)
#20 0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#21 0x00007f9620b3cbea _PyFunction_FastCall (libpython3.6m.so.1.0)
#22 0x00007f9620b7a33e call_function (libpython3.6m.so.1.0)
#23 0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#24 0x00007f9620b0ad9a _PyFunction_FastCall (libpython3.6m.so.1.0)
#25 0x00007f9620b0b3de _PyObject_FastCallDict (libpython3.6m.so.1.0)
#26 0x00007f9620b15261 _PyObject_Call_Prepend (libpython3.6m.so.1.0)
#27 0x00007f9620b0b7fb PyObject_Call (libpython3.6m.so.1.0)
#28 0x00007f9620c12d02 t_bootstrap (libpython3.6m.so.1.0)
#29 0x00007f96207d561b start_thread (libpthread.so.0)
#30 0x00007f961fda691f __clone (libc.so.6)

Stack trace of thread 14072:
#0  0x00007f96207df56b __write (libpthread.so.0)
#1  0x00007f9612aec285 bpf_detach_probe (libbcc.so.0)
#2  0x00007f961742cd1e ffi_call_unix64 (libffi.so.6)
#3  0x00007f961742c68f ffi_call (libffi.so.6)
#4  0x00007f9617640a29 _call_function_pointer (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#5  0x00007f9617640c14 PyCFuncPtr_call (_ctypes.cpython-36m-x86_64-linux-gnu.so)
#6  0x00007f9620b0b204 _PyObject_FastCallDict (libpython3.6m.so.1.0)
#7  0x00007f9620b7a3ad call_function (libpython3.6m.so.1.0)
#8  0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#9  0x00007f9620b3cbea _PyFunction_FastCall (libpython3.6m.so.1.0)
#10 0x00007f9620b7a33e call_function (libpython3.6m.so.1.0)
#11 0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#12 0x00007f9620b3cbea _PyFunction_FastCall (libpython3.6m.so.1.0)
#13 0x00007f9620b7a33e call_function (libpython3.6m.so.1.0)
#14 0x00007f9620ba3eca _PyEval_EvalFrameDefault (libpython3.6m.so.1.0)
#15 0x00007f9620b837e7 _PyEval_EvalCodeWithName (libpython3.6m.so.1.0)
#16 0x00007f9620b84223 function_call (libpython3.6m.so.1.0)
#17 0x00007f9620b0b7fb PyObject_Call (libpython3.6m.so.1.0)
#18 0x00007f9620bfadb1 atexit_callfuncs (libpython3.6m.so.1.0)
#19 0x00007f9620c11f5a call_py_exitfuncs (libpython3.6m.so.1.0)
#20 0x00007f9620c13142 Py_Main (libpython3.6m.so.1.0)
#21 0x000055df9ce42d45 main (python3.6)
#22 0x00007f961fcb000a __libc_start_main (libc.so.6)
#23 0x000055df9ce42eba _start (python3.6)
Process 6993 (python3.6) crashed in perf_reader_event_read()
```

**Comment by yonghong-song** [comment]
@myllynen Followed the installation insn on fc27 and hit an issue as below
```
[yhs@localhost bcc]$ pminfo -f bcc

bcc.proc.io.net.tcp.rx
No value(s) available!

bcc.proc.io.net.tcp.tx
No value(s) available!

bcc.proc.sysfork
    value 87

bcc.disk.all.latency
    inst [0 or "256-511"] value 289
    inst [1 or "512-1023"] value 115
    inst [2 or "2048-4095"] value 21
    inst [3 or "4096-8191"] value 36
    inst [4 or "64-127"] value 11
    inst [5 or "128-255"] value 57
    inst [6 or "1024-2047"] value 13
    inst [7 or "8192-16383"] value 3
[yhs@localhost bcc]$ pmrep bcc.proc.io.perdev
Invalid metric bcc.proc.io.perdev (PM_ERR_NAME Unknown metric name).
[yhs@localhost bcc]$
```

What commands to run in order to repeat the failure? Did you know which bcc tool causing the failure? It is not obvious from the stack trace.

**Comment by myllynen** [comment]
Meh, sorry about leaving out the most essential part!

I'm seeing this typically once in three - five attempts when doing Install/Remove cycle for the PMDA and/or restart pmcd when the PMDA is installed. I don't have a 100% sure recipe but it comes up quite often without trying and when trying to reproduce usually in less than few minutes.

Thanks.

**Comment by yonghong-song** [comment]
Tried consecutive multiple `Install && Remove` cyycles but with no luck:
```
[root@localhost bcc]# pwd
/var/lib/pcp/pmdas/bcc
[root@localhost bcc]# ./Install && ./Remove
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Enabled modules:
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: ['biolatency', 'sysfork', 'tcplife']
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Configuring modules:
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Modules configured.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Initializing modules:
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency: Excluding OS queued time from I/O time.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: sysfork: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: tcplife: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Modules initialized.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Registering metrics:
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Metrics registered.
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Registering helpers:
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21319) Info: Helpers registered.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Enabled modules:
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: ['biolatency', 'sysfork', 'tcplife']
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Configuring modules:
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Modules configured.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Initializing modules:
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency: Excluding OS queued time from I/O time.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: sysfork: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: tcplife: Initialized.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Modules initialized.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Registering metrics:
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: sysfork
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Metrics registered.
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Registering helpers:
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: biolatency
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: tcplife
[Wed Jan 17 17:39:21] pmdabcc(21323) Info: Helpers registered.
You will need to choose an appropriate configuration for installation of
the "bcc" Performance Metrics Domain Agent (PMDA).

  collector	collect performance statistics on this system
  monitor	allow this system to monitor local and/or remote systems
  both		collector and monitor configuration for this system

Please enter c(ollector) or m(onitor) or b(oth) [b] 
Updating the Performance Metrics Name Space (PMNS) ...
Terminate PMDA if already installed ...
Updating the PMCD control file, and notifying PMCD ...
Check bcc metrics have appeared ... 4 metrics and 2 values
Culling the Performance Metrics Name Space ...
bcc ... done
Updating the PMCD control file, and notifying PMCD ...
Check bcc metrics have gone away ... OK
[root@localhost bcc]# 
```
Maybe this is due to my idle FC27 system where really not many stuff going to perf_event_buffer?

**Comment by myllynen** [comment]
Here's I just reproduced this on an up-to-date F27 using the default bcc.conf from my github repo, otherwise the VM is not doing anything at all but perhaps running something like "sync ; echo 3 > /proc/sys/vm/drop_caches ;" that causes disk activity (for biolatency) might help:

1) journalctl -fl in another terminal to spot when the issue happens
2) ./Remove ; echo c | ./Install ; sleep 1 ; pminfo -f bcc ;
3) systemctl restart pmcd

I now managed to reproduce this just after these steps, sometimes repeating 2/3 is needed. Thanks.

**Comment by yonghong-song** [comment]
Yes, I can reproduce the issue. Then I brought up the latest bcc source code and build the binaries/tools and install it. I installed additional llvm/clang packages so that all compiler versions stay at 4.0.1. After this, the `./Remove ; echo c | ./Install ; sleep 1 ; pminfo -f bcc ;` runs correctly, 
```
[root@localhost bcc]# ./Remove ; echo c | ./Install ; sleep 1 ; pminfo -f bcc ; systemctl restart pmcd
Culling the Performance Metrics Name Space ...
bcc ... done
Updating the PMCD control file, and notifying PMCD ...
Check bcc metrics have gone away ... OK
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Enabled modules:
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: ['biolatency', 'sysfork', 'tcplife']
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Configuring modules:
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: sysfork
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: tcplife
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Modules configured.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Initializing modules:
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency: Excluding OS queued time from I/O time.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency: Initialized.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: sysfork
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: sysfork: Initialized.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: tcplife
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: tcplife: Initialized.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Modules initialized.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Registering metrics:
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: sysfork
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: tcplife
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Metrics registered.
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Registering helpers:
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: biolatency
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: tcplife
[Wed Jan 17 19:44:19] pmdabcc(28784) Info: Helpers registered.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Enabled modules:
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: ['biolatency', 'sysfork', 'tcplife']
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Configuring modules:
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: sysfork
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: tcplife
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Modules configured.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Initializing modules:
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency: Excluding OS queued time from I/O time.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency: Initialized.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: sysfork
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: sysfork: Initialized.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: tcplife
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: tcplife: Initialized.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Modules initialized.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Registering metrics:
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: sysfork
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: tcplife
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Metrics registered.
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Registering helpers:
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: biolatency
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: tcplife
[Wed Jan 17 19:44:20] pmdabcc(28795) Info: Helpers registered.
You will need to choose an appropriate configuration for installation of
the "bcc" Performance Metrics Domain Agent (PMDA).

  collector	collect performance statistics on this system
  monitor	allow this system to monitor local and/or remote systems
  both		collector and monitor configuration for this system

Please enter c(ollector) or m(onitor) or b(oth) [b] Updating the Performance Metrics Name Space (PMNS) ...
Terminate PMDA if already installed ...
Updating the PMCD control file, and notifying PMCD ...
Check bcc metrics have appeared ... 1 warnings, 1 metrics and 0 values
bcc: pmLookupName: No PMCD agent for domain of request
[root@localhost bcc]# 
```

I am not able to reproduce the issue any more.

One thing you could help to verify is to install nightly package at
`https://repo.iovisor.org/yum/nightly` to see whether the problem is still there or not.
Thanks.

**Comment by myllynen** [comment]
Somewhat bad news, I switched to the nightly version (I kept using Python 3 but clang/llvm were downgraded (to the versions used earlier with 0.4 packaged for F27) in the process), initially it looked like the issue is gone but in the end I saw it happening again few times.

So perhaps timings have changed somewhere so it's now harder to reproduce but the issue remains.

Thanks.

**Comment by yonghong-song** [comment]
Okay. This is actually not a bcc bug. It is an issue with pcp.
in `modules/tcplife.py`, I added a few comments to illustrate the problem:
```
    def poller(self):
        """ BPF poller """
        try:
            while self.bpf:
                # we have a race here, self.bpf could be gone
                # during kprobe_poll and then program may crash
                # since bpf destructor will free mmap'ed data
                # used by perf event reader
                self.bpf.kprobe_poll()
        except Exception as error: # pylint: disable=broad-except
           self.err(str(error))
           self.err("BPF kprobe poll failed, exiting!")
           if self.bpf:
               self.bpf.cleanup()
           self.bpf = None
```
The issue does not exist in the bcc/tools/tcplife.py. I am not familiar with pcp but if you fix the above race conditions, the problem should be gone.

**Comment by myllynen** [comment]
Thanks for pointing that out, I'm now testing with the below code and I haven't seen crashes any more, please feel free to suggest more elegant ways to handle this if you can think of anything more reliable:

```
        try:
            while True:
                if self.bpf and self.bpf.kprobe_poll():
                    pass
                else:
                    break
        except Exception as error: # pylint: disable=broad-except
           self.err(str(error))
           self.err("BPF kprobe poll failed, exiting!")
           self.bpf = None
```

**Comment by myllynen** [comment]
There was an obvious braino in the above snippet (we ended up to else branch pretty soon thus polling ended), a fixed version is below but I'm seeing crashes with this again:

```
        try:
            while True:
                if self.bpf and self.bpf.kprobe_poll():
                    continue
                else:
                    if not self.bpf:
                        break
        except Exception as error: # pylint: disable=broad-except
            self.err(str(error))
            self.err("BPF kprobe poll failed, exiting!")
            self.bpf = None
```

This polling happens in a background thread so it's different from bcc/tools/tcplife.py which has in the end:

```
while 1:
    b.kprobe_poll()
```

**Comment by yonghong-song** [comment]
The above may not work as BPF object may be gone in the middle of bpf.kprobe_poll().
You probably need something like this:
```
thread 1:
   lock();
   if (self.bpf_exist)   
      self.bpf.kprobe_poll();
   unlock();
thread 2:
   when you are about to destroy bpf object:
    lock();
    self.bpf_exist = false;
    unlock();
```

**Comment by myllynen** [comment]
This is getting interesting.. The locking proposal wouldn't work on idle systems where the first thread is blocked on self.bpf.kprobe_poll() thus the second thread would never get the lock. In the context of PCP BCC PMDA a timeout would occur in the daemon driving the BCC plugin, it'd kill it eventually.

So this brings up the question wouldn't it be most straightforward just to call os._exit() in the atexit handler of the BCC plugin and be done with it, not bothering trying to call self.bpf.cleanup() at all to avoid all these issues.

**Comment by yonghong-song** [comment]
> So this brings up the question wouldn't it be most straightforward just to call os._exit() in the atexit handler of the BCC plugin and be done with it, not bothering trying to call self.bpf.cleanup() at all to avoid all these issues.

We cannot just exit at atexit handler. Our kprobe/uprobe mechanism is based on debugfs and we add some entries in /sys/kernel/debug/tracing/{k,u}probe_events file and we need to do cleanup in these files. @liu-song-6 has implemented file descriptor based {k,u}probe_events tracing which will be available in probably 4.16. Once that becomes main stream, bpf atexit handler could be simplified. 

However, agreed that we could call just enough at atexit handler to cleanup the debugfs system without unmapping the user space memory regions. I tried a patch like this and the pcp still fails. it looks like pcp python program itself does not really exit, and just BPF object itself is gone.

I will implement a simple mechanism to avoid race condition between munmap and read of user space ring buffer. if munmap happens, reading ring buffer will become noop. When reading ring buffer is in action, munmap has to delay.

**Comment by yonghong-song** [comment]
Just submitted a patch to address this issue. https://github.com/iovisor/bcc/pull/1549

**Comment by myllynen** [comment]
Thanks a lot, I've tested the patch locally and it seems to fix the issue.

With additional logging in the poller thread we now see the following in the logs, no more crashes:

pmdabcc(15611) Info: Cleaning up modules:
pmdabcc(15611) Info: tcplife
pmdabcc(15611) Info: tcplife: Poller thread exiting.
pmdabcc(15611) Info: tcplife: BPF/kprobes detached.

I must add that I've seen this one few times (also before the patch), I'm not sure how serious this is (please feel free to close this issue if not significant or if needed I can also file a new issue about this):

pmdabcc(16113) Info: Cleaning up modules:
pmdabcc(16113) Info: tcplife
write(-:kprobes/p_tcp_set_state_bcc_16113): Device or resource busy
pmdabcc(16113) Info: tcplife: BPF/kprobes detached.

Thanks.

**Comment by yonghong-song** [comment]
The error mostly comes from linux/kernel/trace/trace_kprobe.c
```
static int unregister_trace_kprobe(struct trace_kprobe *tk)
{
        /* Enabled event can not be unregistered */
        if (trace_probe_is_enabled(&tk->tp))
                return -EBUSY;

        /* Will fail if probe is being used by ftrace or perf */
        if (unregister_kprobe_event(tk))
                return -EBUSY;

        __unregister_trace_kprobe(tk);
        list_del(&tk->list);

        return 0;
}
```
Not exactly sure what could cause this issue? Could you file a separate issue so we can keep track of this? Thanks!
